armory/utils/metrics.py

```diff
@@ -68,6 +101,127 @@ def l0(x, x_adv):
     return norm(x, x_adv, 0)
 
 
+class MetricList:
+    """
+    Keeps track of all results from a single metric
+    """
+
+    def __init__(self, name, function=None):
+        if function is None:
+            try:
+                self.function = globals()[name]
+            except KeyError:
+                raise KeyError(f"{name} is not part of armory.utils.metrics")
+        elif callable(function):
+            self.function = function
+        else:
+            raise ValueError(f"function must be callable or None, not {function}")
+        self.name = name
+        self._values = []
+
+    def clear(self):
+        self._values.clear()
+
+    def append(self, *args, **kwargs):
+        value = self.function(*args, **kwargs)
+        self._values.extend(value)
+
+    def values(self):
+        return list(self._values)
+
+    def mean(self):
+        return sum(float(x) for x in self._values) / len(self._values)
+
+
+class MetricsLogger:
+    """
+    Uses the set of task and perturbation metrics given to it.
+    """
+
+    def __init__(self, task=None, perturbation=None, means=True, full=False):
+        """
+        task - single metric or list of metrics
+        perturbation - single metric or list of metrics
+        means - whether to return the mean values for each metric
+        full - whether to return the full values for each metric
```

Comment:
I think the word full here is a bit confusing.. especially when read from a config. Maybe something like `record_metric_per_sample` or something?